File: algorithms/medium/check_if_a_parentheses_string_can_be_valid.py
# 
# The intuition for this problem is as follows:
#     1) If the string length is odd, it can never have matched parentheses
#     2) Going from left to right, the count of ')' must never exceed the count of '('
#     3) Going from right to left, the count of '(' must never exceed the count of ')'
# Statements 2) and 3) are true because in the event of such a situation, we will have 
# no opportunity to "fix" the parentheses!
# So the logic is to just traverse left to right and sum all counts of unlocked and '('.
# We choose unlocked because it can be "changed", an unlocked character can be ')'.
# When we encounter a locked ')', we decrement the count and ensure that balance is >= 0.
# Then we apply the same logic going from right to left.
# 

class Solution:
    def canBeValid(self, s: str, locked: str) -> bool:
        if len(s) % 2 == 1:
            return False
        open_count = 0
        for ch, lc in zip(s, locked):
            if lc == '0' or ch == '(':
                open_count += 1
            else:
                open_count -= 1
                if open_count < 0:
                    return False
        #print(f'open_count = {open_count}')
        closed_count = 0
        for ch, lc in zip(reversed(s), reversed(locked)):
            if lc == '0' or ch == ')':
                closed_count += 1
            else:
                closed_count -= 1
                if closed_count < 0:
                    return False
        #print(f'closed_count = {closed_count}')
        return True

# Main section
for s, locked in [
                    ('))()))' , '010100'),
                    ('()()'   , '0000'  ),
                    (')'      , '0'     ),
                    ('))()))' , '011010'),
                    ('())(()(()(()', '100011110110'),
                    ('()', '11'),
                    ('())(()(()(())()())(())((())(()())((())))))(((((((())(()))))(', '100011110110011011010111100111011101111110000101001101001111'),
                 ]:
    print(f's, locked = {s}, {locked}')
    sol = Solution()
    r = sol.canBeValid(s, locked)
    print(f'r = {r}')
    print('====================')

File: algorithms/medium/check_if_a_parentheses_string_can_be_valid.go
package main

import "fmt"

func canBeValid(s string, locked string) bool {
    N := len(s)
    if N % 2 == 1 {
        return false
    }
    openCount := 0
    for i := 0; i < N; i++ {
        ch := s[i]
        lc := locked[i]
        if lc == '0' || ch == '(' {
            openCount++
        } else {
            openCount--
            if openCount < 0 {
                return false
            }
        }
    }
    closedCount := 0
    for i := N - 1; i >= 0; i-- {
        ch := s[i]
        lc := locked[i]
        if lc == '0' || ch == ')' {
            closedCount++
        } else {
            closedCount--
            if closedCount < 0 {
                return false
            }
        }
    }
    return true
}

func main() {
    data := [][]string {
                []string{"))()))"       , "010100"},
                []string{"()()"         , "0000"  },
                []string{")"            , "0"     },
                []string{"))()))"       , "011010"},
                []string{"())(()(()(()" , "100011110110"},
                []string{"()"           , "11"},
                []string{"())(()(()(())()())(())((())(()())((())))))(((((((())(()))))(", "100011110110011011010111100111011101111110000101001101001111"},
            }
    for _, v := range data {
        s, locked := v[0], v[1]
        fmt.Printf("s, locked = %s, %s\n", s, locked)
        r := canBeValid(s, locked)
        fmt.Printf("r = %v\n", r)
        fmt.Println("==================")
    }
}


